File: packages/sdk-core/src/publicState.ts
/**
 * Public (non-secret) pool Merkle state helpers.
 * Clients rebuild the commitment tree from on-chain / mirrored leaves.
 */

import type { PoseidonHasher } from "./note.js";
import { MERKLE_TREE_DEPTH } from "./note.js";
import { buildMerkleTree, getMerklePath, type MerklePath } from "./merkle.js";

export const PUBLIC_STATE_FORMAT = "absolute-privacy-public-state";
export const PUBLIC_STATE_VERSION = 1;

export interface PublicPoolState {
  format: typeof PUBLIC_STATE_FORMAT;
  version: number;
  depth: number;
  /** Commitment leaves in insertion order (decimal strings). */
  commitments: string[];
  /** Poseidon Merkle root for the padded dense tree (decimal string). */
  root: string;
  updatedAt: string;
}

export function createEmptyPublicState(depth: number = MERKLE_TREE_DEPTH): PublicPoolState {
  if (depth <= 0 || depth > 32) {
    throw new Error("depth out of range");
  }
  return {
    format: PUBLIC_STATE_FORMAT,
    version: PUBLIC_STATE_VERSION,
    depth,
    commitments: [],
    root: "0",
    updatedAt: new Date().toISOString(),
  };
}

export function assertPublicPoolState(value: unknown): asserts value is PublicPoolState {
  if (!value || typeof value !== "object") {
    throw new Error("public state must be an object");
  }
  const v = value as Record<string, unknown>;
  if (v.format !== PUBLIC_STATE_FORMAT) {
    throw new Error("unsupported public state format");
  }
  if (v.version !== PUBLIC_STATE_VERSION) {
    throw new Error("unsupported public state version");
  }
  if (!Array.isArray(v.commitments)) {
    throw new Error("public state commitments must be an array");
  }
}

export function parseCommitment(value: string | bigint): bigint {
  if (typeof value === "bigint") return value;
  const s = value.trim();
  if (s.startsWith("0x") || s.startsWith("0X")) {
    return BigInt(s);
  }
  return BigInt(s);
}

/**
 * Rebuild root from commitment leaves (zero-padded to 2^depth).
 * Matches sdk buildMerkleTree / on-chain incremental tree for the same leaf sequence.
 */
export async function recomputePublicStateRoot(
  state: Pick<PublicPoolState, "commitments" | "depth">,
  poseidon: PoseidonHasher
): Promise<bigint> {
  const leaves = state.commitments.map((c) => parseCommitment(c));
  const { root } = await buildMerkleTree(leaves, poseidon, state.depth);
  return root;
}

export async function refreshPublicStateRoot(
  state: PublicPoolState,
  poseidon: PoseidonHasher
): Promise<PublicPoolState> {
  const root = await recomputePublicStateRoot(state, poseidon);
  return {
    ...state,
    root: root.toString(),
    updatedAt: new Date().toISOString(),
  };
}

/**
 * Append a commitment leaf and refresh root. Returns new state + leaf index.
 */
export async function appendCommitment(
  state: PublicPoolState,
  commitment: string | bigint,
  poseidon: PoseidonHasher
): Promise<{ state: PublicPoolState; leafIndex: number }> {
  const width = 1 << state.depth;
  if (state.commitments.length >= width) {
    throw new Error("public state tree is full");
  }
  const next: PublicPoolState = {
    ...state,
    commitments: [...state.commitments, parseCommitment(commitment).toString()],
  };
  const refreshed = await refreshPublicStateRoot(next, poseidon);
  return {
    state: refreshed,
    leafIndex: refreshed.commitments.length - 1,
  };
}

export function findCommitmentIndex(
  state: Pick<PublicPoolState, "commitments">,
  commitment: string | bigint
): number {
  const target = parseCommitment(commitment).toString();
  const idx = state.commitments.findIndex((c) => parseCommitment(c).toString() === target);
  if (idx < 0) {
    throw new Error("commitment not found in public state");
  }
  return idx;
}

export async function merkleWitnessForLeaf(
  state: PublicPoolState,
  leafIndex: number,
  poseidon: PoseidonHasher
): Promise<{ root: bigint; path: MerklePath; leaf: bigint }> {
  const leaves = state.commitments.map((c) => parseCommitment(c));
  if (leafIndex < 0 || leafIndex >= leaves.length) {
    throw new Error("leafIndex out of range for public state");
  }
  const { root, layers, zeros } = await buildMerkleTree(leaves, poseidon, state.depth);
  const path = await getMerklePath(leafIndex, layers, state.depth, zeros);
  return { root, path, leaf: leaves[leafIndex] };
}

File: packages/python-client/absolute_privacy/public_state.py
"""Public pool Merkle state (non-secret)."""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from .merkle import build_merkle_tree, get_merkle_path, MerklePath

PUBLIC_STATE_FORMAT = "absolute-privacy-public-state"
PUBLIC_STATE_VERSION = 1


def _now() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


@dataclass
class PublicPoolState:
    depth: int
    commitments: list[str] = field(default_factory=list)
    root: str = "0"
    updated_at: str = field(default_factory=_now)
    format: str = PUBLIC_STATE_FORMAT
    version: int = PUBLIC_STATE_VERSION
    source: dict[str, Any] | None = None

    def to_json(self) -> dict[str, Any]:
        out: dict[str, Any] = {
            "format": self.format,
            "version": self.version,
            "depth": self.depth,
            "commitments": self.commitments,
            "root": self.root,
            "updatedAt": self.updated_at,
        }
        if self.source is not None:
            out["source"] = self.source
        return out

    @classmethod
    def from_json(cls, data: dict[str, Any]) -> "PublicPoolState":
        if data.get("format") != PUBLIC_STATE_FORMAT:
            raise ValueError("unsupported public state format")
        if int(data.get("version", -1)) != PUBLIC_STATE_VERSION:
            raise ValueError("unsupported public state version")
        return cls(
            depth=int(data["depth"]),
            commitments=[str(c) for c in data.get("commitments") or []],
            root=str(data.get("root") or "0"),
            updated_at=str(data.get("updatedAt") or _now()),
            source=data.get("source"),
        )


def create_empty_public_state(depth: int = 20) -> PublicPoolState:
    if depth <= 0 or depth > 32:
        raise ValueError("depth out of range")
    state = PublicPoolState(depth=depth)
    return refresh_public_state_root(state)


def refresh_public_state_root(state: PublicPoolState) -> PublicPoolState:
    leaves = [int(c) for c in state.commitments]
    root, _ = build_merkle_tree(leaves, state.depth)
    state.root = str(root)
    state.updated_at = _now()
    return state


def append_commitment(state: PublicPoolState, commitment: int | str) -> tuple[PublicPoolState, int]:
    width = 1 << state.depth
    if len(state.commitments) >= width:
        raise ValueError("public state tree is full")
    state.commitments.append(str(int(commitment)))
    refresh_public_state_root(state)
    return state, len(state.commitments) - 1


def find_commitment_index(state: PublicPoolState, commitment: int | str) -> int:
    target = str(int(commitment))
    for i, c in enumerate(state.commitments):
        if str(int(c)) == target:
            return i
    raise ValueError("commitment not found in public state")


def merkle_witness_for_leaf(
    state: PublicPoolState, leaf_index: int
) -> tuple[int, MerklePath, int]:
    leaves = [int(c) for c in state.commitments]
    if leaf_index < 0 or leaf_index >= len(leaves):
        raise ValueError("leafIndex out of range for public state")
    root, layers = build_merkle_tree(leaves, state.depth)
    path = get_merkle_path(leaf_index, layers, state.depth)
    return root, path, leaves[leaf_index]


def load_public_state(path: str | Path) -> PublicPoolState:
    data = json.loads(Path(path).read_text(encoding="utf-8"))
    return PublicPoolState.from_json(data)


def save_public_state(path: str | Path, state: PublicPoolState) -> None:
    Path(path).write_text(json.dumps(state.to_json(), indent=2), encoding="utf-8")
